lib/evmone/instructions.cpp

```diff
@@ -195,190 +196,173 @@ const instruction* opx_beginblock(const instruction* instr, execution_state& sta
+    table[OP_SELFDESTRUCT] = op_selfdestruct;
+
+    table[OP_SHL] = op<shl>;
+    table[OP_SHR] = op<shr>;
+    table[OP_SAR] = op<sar>;
+    table[OP_EXTCODEHASH] = op<extcodehash>;
+    table[OP_CREATE2] = op_create<EVMC_CREATE2>;
+
+    table[OP_CHAINID] = op<chainid>;
+    table[OP_EXTCODEHASH] = op<extcodehash>;
+    table[OP_SELFBALANCE] = op<selfbalance>;
 
-constexpr op_table create_op_table_constantinople() noexcept
-{
-    auto table = create_op_table_byzantium();
-    table[OP_SHL] = {op<shl>, 3, 2, -1};
-    table[OP_SHR] = {op<shr>, 3, 2, -1};
-    table[OP_SAR] = {op<sar>, 3, 2, -1};
-    table[OP_EXTCODEHASH] = {op<extcodehash>, 400, 1, 0};
-    table[OP_CREATE2] = {op_create<EVMC_CREATE2>, 32000, 4, -3};
     return table;
-}
+}();
 
-constexpr op_table create_op_table_istanbul() noexcept
+template <evmc_revision Rev>
+constexpr op_table create_op_table() noexcept
 {
-    auto table = create_op_table_constantinople();
-    table[OP_BALANCE] = {op<balance>, 700, 1, 0};
-    table[OP_CHAINID] = {op<chainid>, 2, 0, 1};
-    table[OP_EXTCODEHASH] = {op<extcodehash>, 700, 1, 0};
-    table[OP_SELFBALANCE] = {op<selfbalance>, 5, 0, 1};
-    table[OP_SLOAD] = {op<sload>, 800, 1, 0};
+    op_table table{};
+    for (size_t i = 0; i < table.size(); ++i)
+    {
+        auto& t = table[i];
+        const auto gas_cost = instr::gas_costs<Rev>[i];
+        if (gas_cost == instr::undefined)
+        {
+            t.fn = op_undefined;
+            t.gas_cost = 0;
+        }
+        else
+        {
+            t.fn = instruction_implementations[i];
+            t.gas_cost = gas_cost;
+            t.stack_req = instr::traits[i].stack_height_required;
+            t.stack_change = instr::traits[i].stack_height_change;
+        }
+    }
     return table;
 }
 
 constexpr op_table op_tables[] = {
-    create_op_table_frontier(),           // Frontier
-    create_op_table_homestead(),          // Homestead
-    create_op_table_tangerine_whistle(),  // Tangerine Whistle
-    create_op_table_tangerine_whistle(),  // Spurious Dragon
-    create_op_table_byzantium(),          // Byzantium
-    create_op_table_constantinople(),     // Constantinople
-    create_op_table_constantinople(),     // Petersburg
-    create_op_table_istanbul(),           // Istanbul
-    create_op_table_istanbul(),           // Berlin
+    create_op_table<EVMC_FRONTIER>(),
+    create_op_table<EVMC_HOMESTEAD>(),
+    create_op_table<EVMC_TANGERINE_WHISTLE>(),
+    create_op_table<EVMC_SPURIOUS_DRAGON>(),
+    create_op_table<EVMC_BYZANTIUM>(),
+    create_op_table<EVMC_CONSTANTINOPLE>(),
+    create_op_table<EVMC_PETERSBURG>(),
+    create_op_table<EVMC_ISTANBUL>(),
+    create_op_table<EVMC_BERLIN>(),
 };
-static_assert(sizeof(op_tables) / sizeof(op_tables[0]) > EVMC_MAX_REVISION,
-    "op table entry missing for an EVMC revision");
+static_assert(
+    std::size(op_tables) > EVMC_MAX_REVISION, "op table entry missing for an EVMC revision");
```

Comment:
maybe could be more strict
```suggestion
    std::size(op_tables) == EVMC_MAX_REVISION + 1, "op table entry missing for an EVMC revision");
```